### ISSUE TITLE
Handle corrupt downloads of Dart SDK better (for ex. a checksum)

### ISSUE THREAD

**Issue by Sun3** [flutter/flutter#21415]
## Steps to Reproduce on Windows 10 64 bit

1. flutter upgrade (the previous version was beta 0.6.0
2. Upgrade fails with errors and unable to run any commands

##Error:
```
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine af42b6dc95bd9f719e43c4e9f29a00640f0f0bba...
Unzipping Dart SDK...
Building flutter tool...
Running pub upgrade...
The system cannot execute the specified program.
Error Unable to 'pub upgrade' flutter tool. Retrying in five seconds...
```

## Logs

```
Not able to run logs
```

```
```

```
not able to run flutter doctor
```

**Comment by DanTup** [comment]
There's a similar report in https://github.com/flutter/flutter/issues/14884 that suggests this may have been blocked by an AntiVirus program - is it possible that's happening here? Are you able to temporarily disable it, or exclude the folder and try again?

**Comment by zoechi** [comment]
There is also #21417

**Comment by Sun3** [comment]
No AntiVirus program. I have been apple to upgrade many times before 0.7.3 

I just reverted to 0.6.0 and works.

**Comment by DanTup** [comment]
Are you able to try manually running running `pub`?

If you change into the directory where you have flutter and run `bin\cache\dart-sdk\bin\pub.bat` does it give the same message?

Can you also confirm whether you're running this from a standard command prompt, PowerShell or a custom shell?

**Comment by Sun3** [comment]
I am running the ```flutter upgrade``` from Windows Command Prompt.

Log of running ```flutter upgrade```

```
C:\Users\sun>flutter channel
Flutter channels:
* beta
  dev
  master

C:\Users\sun>flutter upgrade
Upgrading Flutter from C:\Users\sun\flutter...
From https://github.com/flutter/flutter
   9299c02cf..3b309bda0  beta       -> origin/beta
   3c624f8b2..eab5cd985  dev        -> origin/dev
   8b0227b4a..61719c4e9  master     -> origin/master
 * [new tag]             v0.7.2     -> v0.7.2
 * [new tag]             v0.7.3     -> v0.7.3
 * [new tag]             v0.7.5     -> v0.7.5
 * [new tag]             v0.7.4     -> v0.7.4
Checking out files:  46% (210/456)
Checking out files:  47% (215/456)
Checking out files:  48% (219/456)
Checking out files:  49% (224/456)
Checking out files:  50% (228/456)
Checking out files:  51% (233/456)
Checking out files:  52% (238/456)
Checking out files:  53% (242/456)
Checking out files:  54% (247/456)
Checking out files:  55% (251/456)
Checking out files:  56% (256/456)
Checking out files:  57% (260/456)
Checking out files:  58% (265/456)
Checking out files:  59% (270/456)
Checking out files:  59% (271/456)
Checking out files:  60% (274/456)
Checking out files:  61% (279/456)
Checking out files:  62% (283/456)
Checking out files:  63% (288/456)
Checking out files:  64% (292/456)
Checking out files:  65% (297/456)
Checking out files:  66% (301/456)
Checking out files:  67% (306/456)
Checking out files:  68% (311/456)
Checking out files:  69% (315/456)
Checking out files:  70% (320/456)
Checking out files:  71% (324/456)
Checking out files:  72% (329/456)
Checking out files:  73% (333/456)
Checking out files:  74% (338/456)
Checking out files:  75% (342/456)
Checking out files:  76% (347/456)
Checking out files:  77% (352/456)
Checking out files:  78% (356/456)
Checking out files:  79% (361/456)
Checking out files:  80% (365/456)
Checking out files:  81% (370/456)
Checking out files:  82% (374/456)
Checking out files:  83% (379/456)
Checking out files:  84% (384/456)
Checking out files:  85% (388/456)
Checking out files:  86% (393/456)
Checking out files:  87% (397/456)
Checking out files:  88% (402/456)
Checking out files:  89% (406/456)
Checking out files:  90% (411/456)
Checking out files:  91% (415/456)
Checking out files:  92% (420/456)
Checking out files:  93% (425/456)
Checking out files:  94% (429/456)
Checking out files:  95% (434/456)
Checking out files:  96% (438/456)
Checking out files:  97% (443/456)
Checking out files:  98% (447/456)
Checking out files:  99% (452/456)
Checking out files: 100% (456/456)
Checking out files: 100% (456/456), done.
Updating 9299c02cf..3b309bda0
 .../android_views/android/gradle.properties        |     0
 .../gradle}/wrapper/gradle-wrapper.properties      |     0
 .../lib/client.dart                                |     0
 .../app.tmpl/build.gradle.tmpl                     |     0
 .../androidIdentifier/host/MainActivity.java.tmpl  |     0
 .../src/main/res/drawable/launch_background.xml    |     0
 .../src/main/res/mipmap-hdpi/ic_launcher.png       |   Bin
 .../app.tmpl/src/main/res/values/styles.xml        |     0
 .../settings.gradle.copy.tmpl}                     |     0
 .../{ => library}/Flutter.tmpl/build.gradle.tmpl   |     0
 .../Flutter.tmpl/src/main/AndroidManifest.xml.tmpl |     0
 .../src/main/java/io/flutter/facade/Flutter.java   |     0
 .../java/io/flutter/facade/FlutterFragment.java    |     0
 .../include_flutter.groovy.copy.tmpl}              |     0
 440 files changed, 19639 insertions(+), 7159 deletions(-)

Upgrading engine...
Checking Dart SDK version...
Downloading Dart SDK from Flutter engine af42b6dc95bd9f719e43c4e9f29a00640f0f0bba...
Unzipping Dart SDK...
Building flutter tool...
Running pub upgrade...
The system cannot execute the specified program.
Error Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries left)
```

Running pub.bat from the folder c:\Users\sun\flutter\bin\cache\dart-sdk\bin\

```
'"C:\Users\sun\flutter\bin\cache\dart-sdk\bin\dart"' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by Sun3** [comment]
ps: pub.bat file contents

```
@echo off
REM Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
REM for details. All rights reserved. Use of this source code is governed by a
REM BSD-style license that can be found in the LICENSE file.

setlocal
rem Handle the case where dart-sdk/bin has been symlinked to.
set DIR_NAME_WITH_SLASH=%~dp0
set DIR_NAME=%DIR_NAME_WITH_SLASH:~0,-1%%
call :follow_links "%DIR_NAME%", RETURNED_BIN_DIR
rem Get rid of surrounding quotes.
for %%i in ("%RETURNED_BIN_DIR%") do set BIN_DIR=%%~fi

rem Get absolute full name for SDK_DIR.
for %%i in ("%BIN_DIR%\..\") do set SDK_DIR=%%~fi

rem Remove trailing backslash if there is one
IF %SDK_DIR:~-1%==\ set SDK_DIR=%SDK_DIR:~0,-1%

set VM_OPTIONS=
set USING_DART_1=

rem We allow extra vm options to be passed in through an environment variable.
if not "_%DART_VM_OPTIONS%_" == "__" (
  set VM_OPTIONS=%VM_OPTIONS% %DART_VM_OPTIONS%
  for %%o in (%DART_VM_OPTIONS%) do (
    if "%%o" equ "--no-preview-dart-2" set USING_DART_1=y
  )
)

if defined USING_DART_1 (
  echo "Pub no longer supports Dart 1"
) else (
  "%BIN_DIR%\dart" %VM_OPTIONS% "%BIN_DIR%\snapshots\pub.dart.snapshot" %*
)

endlocal

exit /b %errorlevel%

:follow_links
setlocal
for %%i in (%1) do set result=%%~fi
set current=
for /f "usebackq tokens=2 delims=[]" %%i in (`dir /a:l "%~dp1" 2^>nul ^
                                             ^| find ">     %~n1 [" 2^>nul`) do (
  set current=%%i
)
if not "%current%"=="" call :follow_links "%current%", result
endlocal & set %~2=%result%
goto :eof

:end

```

**Comment by Sun3** [comment]
ps: FYI in the ```C:\Users\sun\flutter\bin\cache\dart-sdk\bin\dart``` folder is see a ```dart.lib``` but not a ```dart.exe```.

**Comment by DanTup** [comment]
> FYI in the C:\Users\sun\flutter\bin\cache\dart-sdk\bin\dart folder is see a dart.lib but not a dart.exe.

It should be directly in the `bin` folder, not `bin\dart`. Do you see it there?

**Comment by Sun3** [comment]
Sorry, I mistyped the folder I meant the ```C:\Users\sun\flutter\bin\cache\dart-sdk\bin``` and the ```dart.exe``` is missing but I can see the ```dart.lib``` file.

**Comment by DanTup** [comment]
@Sun3 Can you still reproduce? It'd be interesting to inspect the zip file that was extracted to see whether it has a dart.exe in. The code that does the extracting is here:

https://github.com/flutter/flutter/blob/154dbc597544b60bc4b090c1f85e828e7709c8d0/bin/internal/update_dart_sdk.ps1#L56

**Comment by Sun3** [comment]
Every time I try to update from v0.6.0 to v07.3 I received the same errors in my Windows 10 x64 box. On my Mac's they all upgrade fine.

I have placed below the Contents of  update_dart_sdk.ps1 and I cannot find the dart.exe file extracted.

But while downloading the dart-sdk-windows-x64.zip file I made a copy before it was automatically removed and the zip file does contain the dart.exe file. I figured let me unzip the file manually but when I do not everything is getting unzipped. Very interesting.

Contents of update_dart_sdk.ps1
```
# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.


# ---------------------------------- NOTE ---------------------------------- #
#
# Please keep the logic in this file consistent with the logic in the
# `update_dart_sdk.sh` script in the same directory to ensure that Flutter
# continues to work across all platforms!
#
# -------------------------------------------------------------------------- #

$ErrorActionPreference = "Stop"

$progName = Split-Path -parent $MyInvocation.MyCommand.Definition
$flutterRoot = (Get-Item $progName).parent.parent.FullName

$cachePath = "$flutterRoot\bin\cache"
$dartSdkPath = "$cachePath\dart-sdk"
$engineStamp = "$cachePath\engine-dart-sdk.stamp"
$engineVersion = (Get-Content "$flutterRoot\bin\internal\engine.version")

$oldDartSdkPrefix = "dart-sdk.old"

if ((Test-Path $engineStamp) -and ($engineVersion -eq (Get-Content $engineStamp))) {
    return
}

Write-Host "Downloading Dart SDK from Flutter engine $engineVersion..."
$dartSdkBaseUrl = $Env:FLUTTER_STORAGE_BASE_URL
if (-not $dartSdkBaseUrl) {
    $dartSdkBaseUrl = "https://storage.googleapis.com"
}
$dartZipName = "dart-sdk-windows-x64.zip"
$dartSdkUrl = "$dartSdkBaseUrl/flutter_infra/flutter/$engineVersion/$dartZipName"

if (Test-Path $dartSdkPath) {
    # Move old SDK to a new location instead of deleting it in case it is still in use (e.g. by IntelliJ).
    $oldDartSdkSuffix = 1
    while (Test-Path "$cachePath\$oldDartSdkPrefix$oldDartSdkSuffix") { $oldDartSdkSuffix++ }
    Rename-Item $dartSdkPath "$oldDartSdkPrefix$oldDartSdkSuffix"
}
New-Item $dartSdkPath -force -type directory | Out-Null
$dartSdkZip = "$cachePath\$dartZipName"

Try {
    Import-Module BitsTransfer
    Start-BitsTransfer -Source $dartSdkUrl -Destination $dartSdkZip
}
Catch {
    Write-Host "Downloading the Dart SDK using the BITS service failed, retrying with WebRequest..."
    Invoke-WebRequest -Uri $dartSdkUrl -OutFile $dartSdkZip
}

Write-Host "Unzipping Dart SDK..."
If (Get-Command 7z -errorAction SilentlyContinue) {
    # The built-in unzippers are painfully slow. Use 7-Zip, if available.
    & 7z x $dartSdkZip "-o$cachePath" -bd | Out-Null
} ElseIf (Get-Command 7za -errorAction SilentlyContinue) {
    # Use 7-Zip's standalone version 7za.exe, if available.
    & 7za x $dartSdkZip "-o$cachePath" -bd | Out-Null
} ElseIf (Get-Command Expand-Archive -errorAction SilentlyContinue) {
    # Use PowerShell's built-in unzipper, if available (requires PowerShell 5+).
    Expand-Archive $dartSdkZip -DestinationPath $cachePath
} Else {
    # As last resort: fall back to the Windows GUI.
    $shell = New-Object -com shell.application
    $zip = $shell.NameSpace($dartSdkZip)
    foreach($item in $zip.items()) {
        $shell.Namespace($cachePath).copyhere($item)
    }
}

Remove-Item $dartSdkZip
$engineVersion | Out-File $engineStamp -Encoding ASCII

# Try to delete all old SDKs.
Get-ChildItem -Path $cachePath | Where {$_.BaseName.StartsWith($oldDartSdkPrefix)} | Remove-Item -Recurse -ErrorAction SilentlyContinue
```

**Comment by DanTup** [comment]
@Sun3 You could add some debug output (`Write-Output "Using Expand-Archive"` etc.) to that file to figure out which unzip mechanism it's using (eg. `7z`/`7za`/`Expand-Archive`) and then try each manually from the terminal to see if that makes any difference.

If you do have any antivirus software or similar, it may be worth trying to temporarily disable it to see if that makes any difference regardless of whether it's reporting that it's blocking anything.

**Comment by Sun3** [comment]
Not able to duplicate anymore. I manually downloaded the v0.7.3 zip file from flutter.io and deleted current flutter folder. Then I unzipped it, and I can run flutter upgrade, and any flutter commands again. I made no changes to the computer. I had done this step 3-4 times before it worked.

**Comment by DanTup** [comment]
Very strange. Oh well, thanks for confirming and if you do see it again and can try the steps above, please do!

**Comment by Hixie** [comment]
We might want to consider adding some sort of checksum testing or something when downloading these files, we've seen this kind of corruption-during-download before.

**Comment by DanTup** [comment]
Ah, I assumed something was going bad during the extraction, but if it might've been a corrupt zip then validating it would make sense. Do we already have an access to some checksum, or would it require a change to whatever is building/hosting the Dart SDKs?

**Comment by Hixie** [comment]
I don't think we have any relevant logic in place yet.

**Comment by DanTup** [comment]
Ok, makes sense to have an open issue about it.

**Comment by Sun3** [comment]
ps: On the next Beta upgrade, if I have further issues I will post an update.

**Comment by DanTup** [comment]
If you do see it again, it's worth grabbing a copy of the zip file that has the issues to verify if/how it was corrupt.

I'm not sure why it seemed consistently broken in the latest but fine in the previous, it could just be fluke. I think I may have had a corrupt file once when I opened VS Code which started `flutter daemon` which kicked off a download but I closed VS Code before it was finished. I thought it noticed this and started again though (it output a message like it was trying to resume the download, but failed, but I don't remember the details since it was some time ago).
